Bluetooth: Host: bsim compilation improvements

ardo-nordic opened the pull request:
- All compile.sh files in bsim/host removed and replaced with testcase.yaml files which are used for building with twister
- Some host tests which compile and run on bsim53 (net core) are enabled
- Some conf files streamlined to avoid code duplication
- CI bsim files updated correspondingly

Fixes #77055

Thalley commented:
Just out of curiosity, do you have an example of what the user should do now to manually test a specific test? 

Previously I called the `compile.sh` and the ran one of the test scripts to run a specific test locally. What should a user to do accomplish the same with this change?

aescolar reviewed (dismissed):
Thanks @ardo-nordic 
A few minor comments

ardo-nordic commented:
@Thalley Build is now a simple `west twister -T <path_to_test> -p <platform>` command.

ardo-nordic commented on tests/bsim/bluetooth/host/_tests.nrf5340bsim_nrf5340_cpunet.sh:
Allows running all supported nrf53 tests. The problem as it's implemented now with bsim harness in twister only allowing to build tests, is that shell scripts which are used to run them have no idea which tests are supported on which platform. Everything is supported on nrf52_bsim, so that's not an issue, but in their current form, few tests can be build and run on nrf53 with bsim. Actually in host none are run now, but in this commit I'm enabling those which are working without any additional efforts.

Thalley commented on tests/bsim/bluetooth/host/adv/extended/testcase.yaml:
```diff
@@ -0,0 +1,20 @@
+  build_only: true
+  tags:
+    - bluetooth
+  platform_allow:
+    - nrf52_bsim
+    - nrf5340bsim/nrf5340/cpunet
+  harness: bsim
```
It would have been nice to see at least 1 test for the nRF5340 cpuapp as well, to act as a reference for how to use this with sysbuild/dual CPU

Thalley commented on tests/bsim/bluetooth/host/adv/encrypted/css_sample_data/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+    tags:
+      - bluetooth
+    platform_allow:
+      - nrf52_bsim
```
@aescolar should this also be?

```suggestion
      - nrf52_bsim/native
```

aescolar replied:
It does not matter as it is the only option.

Thalley commented on tests/bsim/bluetooth/host/adv/chain/testcase.yaml:
```diff
@@ -0,0 +1,11 @@
+tests:
+  bluetooth.host.adv.chain:
+    build_only: true
```
> build_only: true

Is it planned that these will also eventually run the tests?

aescolar replied:
:)
step 1) (what we already have) was to be able build and place the executable in the "right" place.
step 2) is to have dependencies between tests and images (so twister knows what images a test needs)
step 3) would be to also be able to run them from twister.

ardo-nordic replied:
bsim harness in twister as of now only supports building, and it also copies files to appropriate location. To run them, old shell scripts will still be used.

Thalley commented on tests/bsim/bluetooth/tests.nrf5340bsim_nrf5340_cpunet.txt:
```diff
@@ -2,3 +2,16 @@
 # built in the net core)
 # This file is used in CI to select which tests are run
 tests/bsim/bluetooth/ll/
+tests/bsim/bluetooth/host/att/eatt_notif
+tests/bsim/bluetooth/host/misc/disable
+tests/bsim/bluetooth/host/misc/unregister_conn_cb
+tests/bsim/bluetooth/host/adv/periodic
+tests/bsim/bluetooth/host/adv/extended
+tests/bsim/bluetooth/host/adv/chain
+tests/bsim/bluetooth/host/l2cap/send_on_connect
+tests/bsim/bluetooth/host/central
+tests/bsim/bluetooth/host/privacy/peripheral
+tests/bsim/bluetooth/host/privacy/central
+tests/bsim/bluetooth/host/gatt/authorization
+tests/bsim/bluetooth/host/gatt/general
```
What do we gain from running these on the cpunet?
I think it's OK to add one or two just to do a smoke test, but I'm not sure we would see any significant change in behavior between the nrf52bsim and nrf5340 cpunet, and these would probably have been better suited for the cpuapp where there's a more noticeable difference in behavior.

aescolar commented on tests/bsim/bluetooth/host/adv/compile.sh:
The removal of these .sh files should be in the same commit that changes the call from tests/bsim/bluetooth/host/compile.sh  (otherwise with this commit alone things don't work)

aescolar commented on tests/bsim/bluetooth/compile.nrf5340bsim_nrf5340_cpunet.sh:
```diff
@@ -13,5 +13,6 @@ export BOARD="${BOARD:-nrf5340bsim/nrf5340/cpunet}"
 source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
 
 ${ZEPHYR_BASE}/tests/bsim/bluetooth/ll/compile.sh
+${ZEPHYR_BASE}/tests/bsim/bluetooth/host/compile.sh
```
Nit: I would still place this coverage increase change in its own commit.

aescolar reviewed (commented):
Almost there from my point of view

ardo-nordic committed:
tests: bsim: bluetooth: host: use twister to compile tests

Bluetooth host bsim tests are compiled with shell scripts, but it is
possible to use twister, which supports compilation (but not running
at the moment). Compiling with twister requires fewer shell scripts,
provides cleaner interface and more flexible configuration.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: bluetooth: host: streamline conf files

Currently different test cases within the same folder are using largely the
same config files with only subtle variations. It is better instead to have
a base config file with general configuration, aided by overlay files for
specific cases.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: bluetooth: host: enable on nrf5340bsim/nrf5340/cpunet

Some bluetooth host tests can be run on nrf5340bsim/nrf5340/cpunet.
This commit enables building and running these tests.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

ardo-nordic committed:
tests: bsim: bluetooth: using full nrf52bsim board name

Some places kept using partial nrf52bsim board name which led to
inconsistencies.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

aescolar commented on tests/bsim/bluetooth/host/gatt/notify_multiple/test_scripts/_notify-debug.sh:
```diff
@@ -15,6 +15,7 @@ source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
 
 simulation_id="notify_multiple"
 verbosity_level=2
+BOARD="${BOARD:-nrf52_bsim/native}"
```
This line was just removed in a previous commit. Seems this is a bad rebase merge(?)

aescolar commented on tests/bsim/bluetooth/ll/_compile_permutate_kconfigs.sh:
```diff
@@ -14,8 +14,9 @@ DEBUG_PERMUTATE=false
 
 source ${ZEPHYR_BASE}/tests/bsim/compile.source
-
+BOARD="${BOARD:-nrf52_bsim/native}"
```
These 2 lines where just removed in another commit.
Seems these 2 lines were just added back in a bad merge?

aescolar commented on tests/bsim/bluetooth/tests.nrf5340bsim_nrf5340_cpunet.txt:
```diff
@@ -10,7 +10,6 @@ tests/bsim/bluetooth/host/adv/extended
 tests/bsim/bluetooth/host/adv/chain
 tests/bsim/bluetooth/host/l2cap/send_on_connect
 tests/bsim/bluetooth/host/central
-tests/bsim/bluetooth/host/privacy/peripheral
```
Please just squash this commit into the commit in which you just added it ("tests: bsim: bluetooth: host: enable on nrf5340bsim/nrf5340/cpunet")

cvinayak commented on tests/bsim/bluetooth/host/privacy/peripheral/prj_rpa_expired.conf:
Rename to overlay-rpa_expired.conf

cvinayak commented on tests/bsim/bluetooth/host/privacy/peripheral/prj_rpa_sharing.conf:
Rename to overlay-rpa_sharing.conf

cvinayak commented on tests/bsim/bluetooth/host/privacy/peripheral/testcase.yaml:
```diff
@@ -0,0 +1,22 @@
+    harness_config:
+      bsim_exe_name: tests_bsim_bluetooth_host_privacy_peripheral_prj_conf
+  bluetooth.host.privacy.peripheral_rpa_expired:
+    harness_config:
+      bsim_exe_name: tests_bsim_bluetooth_host_privacy_peripheral_prj_rpa_expired_conf
+    extra_args:
+      EXTRA_CONF_FILE=prj_rpa_expired.conf
+  bluetooth.host.privacy.peripheral_rpa_sharing:
+    harness_config:
+      bsim_exe_name: tests_bsim_bluetooth_host_privacy_peripheral_prj_rpa_sharing_conf
+    extra_args:
+      EXTRA_CONF_FILE=prj_rpa_sharing.conf
```
```suggestion
    harness_config:
      bsim_exe_name: tests_bsim_bluetooth_host_privacy_peripheral_prj_conf_overlay_rpa_expired_conf
    extra_args:
      EXTRA_CONF_FILE=overlay-rpa_expired.conf
  bluetooth.host.privacy.peripheral_rpa_sharing:
    harness_config:
      bsim_exe_name: tests_bsim_bluetooth_host_privacy_peripheral_prj_conf_overlay_rpa_sharing_conf
    extra_args:
      EXTRA_CONF_FILE=overlay-rpa_sharing.conf
```

cvinayak commented on tests/bsim/bluetooth/host/security/ccc_update/prj_2.conf:
Rename to overlay-2.conf

cvinayak commented on tests/bsim/bluetooth/host/security/ccc_update/testcase.yaml:
```diff
@@ -0,0 +1,17 @@
+    harness_config:
+      bsim_exe_name: tests_bsim_bluetooth_host_security_ccc_update_prj_conf
+  bluetooth.host.security.ccc_update_2:
+    harness_config:
+      bsim_exe_name: tests_bsim_bluetooth_host_security_ccc_update_prj_2_conf
+    extra_args:
+      EXTRA_CONF_FILE=prj_2.conf
```
```suggestion
    harness_config:
      bsim_exe_name: tests_bsim_bluetooth_host_security_ccc_update_prj_conf_overlay_2_conf
    extra_args:
      EXTRA_CONF_FILE=overlay-2.conf
```

cvinayak reviewed (dismissed):
Very good improvement. Requested changes to reflect the explicit use of extra/overlay conf files in builds.

ardo-nordic commented:
Do we really need this change? File names are already long enough, and it's easy to match what belongs where looking at testcase.yaml file, so unless anyone is really taking into account the exact binary file names and use it for something, I'd just leave things where they are.

cvinayak commented:
> Do we really need this change? File names are already long enough, and it's easy to match what belongs where looking at testcase.yaml file, so unless anyone is really taking into account the exact binary file names and use it for something, I'd just leave things where they are.

It is easy with how it is before (and currently for the ones with no change), the lines indicate which ones overlay/extra variant failed to build without having to browse to read the testcase.yaml :
![image](https://github.com/user-attachments/assets/0afc360a-5d2c-4444-98da-42b97f4bf1f5)

of course, the test scripts are still very cryptic today.

Since you deleted files etc.. in this PR, yes, having the renaming to be consistent with existing exename makes sense in this PR.

ardo-nordic commented:
With twister however, when failing to build, it will print the test name as specified in testcase.yaml, so analyzing the failure would require looking into yaml file anyway. Which in my opinion, is a good thing, as this file is way more readable and gives a much better overview of the overall test configuration, rather parsing extra long file executable name and trying to deduce all the info from it. So I don't think the naming convention is extremely relevant in that case.

cvinayak commented:
> So I don't think the naming convention is extremely relevant in that case.

It is revelant as you have changed how the tests where built using `CONF_FILE`, now to explicitly merge as extra config files due to migration for using `EXTRA_CONF_FILE`. The scope of the tests have changed and there can be false positives in the tests passing in this PR CI. Have we checked all the `zephyr/.config` files to be same as before the PR changes?

ardo-nordic commented:
> > So I don't think the naming convention is extremely relevant in that case.
> 
> It is revelant as you have changed how the tests where built using `CONF_FILE`, now to explicitly merge as extra config files due to migration for using `EXTRA_CONF_FILE`. The scope of the tests have changed and there can be false positives in the tests passing in this PR CI. Have we checked all the `zephyr/.config` files to be same as before the PR changes?

Well, in all the comments above, we were speaking only about naming convention, haven't we? Naming convention isn't going to prevent errors by itself. My point is with this change, testcase.yaml file provide a full overview of test configuration, including which overlay (or extra config) files  are used for which test. Having it in place, you don't need to rely on very precise (and very long) binary file name to understand for which particular configuration it was build.

As for the change in the conf files done in that PR, only changed it in places where there were obvious common set of config parameters, and config files that deviated from the baseline by only one or few config params were changed to be extra config files, to avoid code duplication.

ardo-nordic committed:
scripts: twister: adding default cmake/config arguments for bsim

Bsim tests need some addtional kconfig arguments enabled by default.

Signed-off-by: Artur Dobrynin <artur.dobrynin@nordicsemi.no>

Thalley commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -651,6 +651,13 @@ def run_cmake(self, args="", filter_stages=None):
+        if self.instance.testsuite.harness == 'bsim':
+            cmake_args.extend([
+                '-DCMAKE_EXPORT_COMPILE_COMMANDS=ON',
+                '-DCONFIG_ASSERT=y',
```
If you enable this by default, should we remove the `CONFIG_ASSERT=y` from the config files of the affected tests?

ardo-nordic replied:
We can, but it's always been enabled in compile.source, but still present in conf files, probably just copy-pasted over from one to another, or may be it was added to compile source later.

aescolar commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -651,6 +651,13 @@ def run_cmake(self, args="", filter_stages=None):
         ]
 
+        if self.instance.testsuite.harness == 'bsim':
+            cmake_args.extend([
```
@ardo-nordic
How come did you add this here? (just because they were added in compile.sh ? or did somebody ask for them?

ardo-nordic replied:
Same arguments are added in zephyr/tests/bsim/compile.source by default.

aescolar replied:
For context: I'm the one who added run_parallel and compile.sh, and I'm the one who added setting COVERAGE=y by default. I did so because at the time we were generating coverage reports from PRs.
But we haven't done so for a while, and beyond telling people how to generate them in the documentation, I don't think many people is using it for anything nowadays.
https://docs.zephyrproject.org/latest/develop/test/coverage.html
https://docs.zephyrproject.org/latest/develop/test/bsim.html#test-coverage-and-babblesim
So I have been feeling that building (and running) with coverage enabled has been a waste of CI time, and have been inclined to remove it for quite a while now. twister has it's own "--coverage" flag we instruct people to use if they want to generate a report.

Wrt ASSERTS I think we should have that set in the tests' conf files, or maybe in the board defconfig instead (@jori-nordic had set it in the compile.source in 1b4844aad0e0e49422b7f0a0ce92001113431319) (@Thalley @jhedberg @alwa-nordic feel free to chip in). I mean, I think it is good to set it for testing, I just don't think it is good to set it in hidden places.

The CMAKE_EXPORT_COMPILE_COMMANDS came bundled in
https://github.com/zephyrproject-rtos/zephyr/commit/f023b5f611df8f8fa975bf2ebd3e80239b4e9278
from the llcp refactoring topic branch. I don't think we ever really needed that set(?) (@kruithofa do you recall why this was done?)

Thalley replied:
We should have asserts enabled by default for all bsim testing IMO. If there is a better way to set it by default (and not per test), then I'm OK with moving it.

aescolar replied:
> We should have asserts enabled by default for all bsim testing IMO.

@Thalley With all bsim testing, are you thinking: a) "whenever we build targeting this board" or b) "whenever we run any sample or test in CI", c) "whenever we build any test (not sample) for this board", d) something else 
?
